### PR TITLE
Mean surface reflectance image

### PR DESCRIPTION
**Pull request by uchchwhash**
- Produce mean surface reflectance image as part of sensitivity analysis

In order to reduce disk space usage, the calculation of the sum and the count of pixels is organized into a binary tree of tasks. Once that info is collected, the source datasets are immediately cleaned up.

### CODE REVIEW

File: tesp/sensitivity.py [hunk 1]
```diff
@@ -32,33 +37,410 @@ class ExperimentList(luigi.WrapperTask):
...
+
+    def add_image(src_affine, src_product):
+        """
+        Add this image to the target array.
+        """
+        src_ds = find_dataset_by_name(src_product, band_name)
+
+        # find the target window
+        ul = ~target_affine * (src_affine * (0, 0))
+        lr = ~target_affine * (src_affine * src_ds.shape)
```
Comment (sixy6e) [inline tesp/sensitivity.py]: Careful with this line
src_affine * src_ds.shape  looks incorrect.  src_ds.shape will return (y, x) , but requires affine * (x, y)

If your shape is square then ok, but probably safer to not be specific

Comment (uchchwhash) [reply tesp/sensitivity.py]: Thanks! This was it! Seems to work now, asking for your review again.

This time there is a class called `GeoBox` which may someday get merged back into `GriddedGeoBox`.

**Comment by sixy6e** [comment]
Thanks for all the comments @jeremyh it's great to have someone that hasn't touched the code base before take the time to give a thorough review of this PR.

@uchchwhash we'll address @jeremyh comments and suggestions before merging

File: tesp/sensitivity.py [hunk 2]
```diff
@@ -106,8 +203,348 @@ def run(self):
...
+
+
+def merge_datasets(left_product, left_mask, right_product, right_mask, target_product, band_name):
+    """
+    Merge images from two `.h5` groups in two different `.h5` container.
+    """
+    left_ds = find_dataset_by_name(left_product, band_name)
+    right_ds = find_dataset_by_name(right_product, band_name)
+
+    assert left_ds.attrs['crs_wkt'] == right_ds.attrs['crs_wkt'], "I can't merge images from different CRSs yet"
```
[the remark this answers is not in the document]
Comment (uchchwhash) [reply tesp/sensitivity.py]: Fixed.

File: tesp/sensitivity.py [hunk 3]
```diff
@@ -106,8 +203,348 @@ def run(self):
...
+        src_data, count_data = sum_and_count(src_product, src_mask, band_name)
+
+        # add the data to appropriate window in the target array
+        target_ds[window] += src_data
+        target_count[window] += count_data
+
+    add_image(left_box, left_product, left_mask)
+    add_image(right_box, right_product, right_mask)
+
+    # copy over metadata
```
[the remark this answers is not in the document]
Comment (uchchwhash) [reply tesp/sensitivity.py]: Yes, that is the intention. We only care about the `alias` and the `crs_wkt` though. These two are common to the two datasets.

File: tesp/sensitivity.py [hunk 4]
```diff
@@ -106,8 +203,348 @@ def run(self):
...
+    for key in left_ds.attrs:
+        if key in right_ds.attrs:
+            try:
+                if left_ds.attrs[key] == right_ds.attrs[key]:
+                    target_ds.attrs[key] = left_ds.attrs[key]
+                    target_count.attrs[key] = left_ds.attrs[key]
+            except ValueError:
```
[the remark this answers is not in the document]
Comment (uchchwhash) [reply tesp/sensitivity.py]: That's a great idea.

**Comment by uchchwhash** [comment]
@jeremyh yes the `luigi` scheduler hands out the tasks to a predefined number of workers (28 in our case), so all the leaf tasks would execute at the same time. Periodically the `luigi` scheduler prunes the task tree, figures out that the dependencies for some tasks are all satisfied, and runs them.

**Pull request merged**